llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp

```diff
@@ -0,0 +1,73 @@
+//===-- MachineFunctionHashBuilder.cpp ----------------------------------*-
+// C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the implementation of pass about calculating machine
+/// function hash.
+//===----------------------------------------------------------------------===//
+#include "llvm/CodeGen/MachineFunctionHashBuilder.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/MD5.h"
+#include "llvm/Support/raw_ostream.h"
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+
+using namespace llvm;
+
+// Calculate machine function hash in level order traversal.
+// The control flow graph is uniquely represented by its level-order traversal.
+static uint64_t calculateMBBCFGHash(const MachineFunction &MF) {
+  if (MF.empty())
+    return 0;
+  std::unordered_set<const MachineBasicBlock *> Visited;
+  MD5 Hash;
+  std::queue<const MachineBasicBlock *> WorkList;
+  WorkList.push(&*MF.begin());
+  while (!WorkList.empty()) {
+    const MachineBasicBlock *CurrentBB = WorkList.front();
+    WorkList.pop();
+    uint32_t Value = support::endian::byte_swap<uint32_t, endianness::little>(
+        CurrentBB->getBBID()->BaseID);
+    uint32_t Size = support::endian::byte_swap<uint32_t, endianness::little>(
+        CurrentBB->succ_size());
+    Hash.update(ArrayRef((uint8_t *)&Value, sizeof(Value)));
```

Comment:
You may update the hash several times with the same element. A single item may be inserted and visited several times since `Visited.count(CurrentBB)` is checked later. I think this is unnecessary. You can do this check before pushing items in the worklist. Also, please initialize `Visited` with the entry block.

Reply:
For example, the cfg is:
`0->2->3`
`  ->1->3 `
I want to restore the cfg by BB's id and BB's successor's size. So the sequence will be 
`0(id) 2(successor size) 1(id) 1(successor size) 2(id) 1(successor size) 3(id) 0(successor size) 3(id) 0(successor size)`.

Reply:
I don't fully understand the various cases you are trying to tolerate with this hashing.  You are only hashing the size and the id of the block.   What are the cases you are trying to capture here?  Why not hash the contents of the block?

In this mechanism, the cfg hashing will not tolerate a change to the compiler where the blocks had the same CFG but somehow got re-numbered, right?

Reply:
> In this mechanism, the cfg hashing will not tolerate a change to the compiler where the blocks had the same CFG but somehow got re-numbered, right?

Thanks for reply. Yes, you are right.
I think we can divide cfg hashing to three version: loose hash, strict hash, full hash which Bolt do. For example, how about the strict hash is calculated by the size and the id of the block and the full hash is calculated by the contents of the function? More strict hash will cause more compile time.

Reply:
Can you measure the additional overhead of doing a full/strict hash over a loose hash?